<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 3.1.1">
  <meta name="generator" content="Hugo 0.51" />
  <meta name="author" content="T. Jeffrey Cole">

  
  
  
  
    
  
  <meta name="description" content="Tutorial for bioinformatics pipelines">

  
  <link rel="alternate" hreflang="en-us" href="/post/reproducibilitytutorial/">

  


  

  

  

  
  
  
  <meta name="theme-color" content="hsl(159, 97%, 41%)">
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.4.1/css/all.css" integrity="sha384-5sAR7xN1Nv6T6+dT2mhtzEpVJvfS3NScPQTrOxhwjIuvcA67KV2R5Jz6kr4abQsz" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/dracula.min.css" crossorigin="anonymous">
        
      
    

    

    

  

  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700|Roboto:400,400italic,700|Roboto+Mono">
  

  <link rel="stylesheet" href="/styles.css">
  

  
  
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="tijeco">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="tijeco">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/reproducibilitytutorial/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@wandering_jeff">
  <meta property="twitter:creator" content="@wandering_jeff">
  
  <meta property="og:site_name" content="tijeco">
  <meta property="og:url" content="/post/reproducibilitytutorial/">
  <meta property="og:title" content="Reproducubility Tutorial | tijeco">
  <meta property="og:description" content="Tutorial for bioinformatics pipelines">
  
  
    
  <meta property="og:image" content="/img/icon-192.png">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2017-04-18T00:00:00-04:00">
  
  <meta property="article:modified_time" content="2017-04-18T00:00:00-04:00">
  

  

  

  <title>Reproducubility Tutorial | tijeco</title>

</head>
<body id="top" data-spy="scroll" data-target="#TableOfContents" data-offset="71" class="dark">
  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header fixed-top">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


<nav class="navbar navbar-light fixed-top navbar-expand-lg py-0" id="navbar-main">
  <div class="container">

    
      <a class="navbar-brand" href="/">tijeco</a>
      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
        <span><i class="fas fa-bars"></i></span>
      </button>
      

    
    <div class="collapse navbar-collapse" id="navbar">

      
      
      <ul class="navbar-nav ml-auto">
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        <li class="nav-item dropdown">
          <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-haspopup="true">
            
            <span>Projects</span>
            
            <span class="caret"></span>
          </a>
          <ul class="dropdown-menu">
            
            <li class="dropdown-item my-0 py-0 mx-0 px-0">
              <a href="/tags/bioinformatics">
                
                <span>bioinformatics</span>
                
              </a>
            </li>
            
            <li class="dropdown-item my-0 py-0 mx-0 px-0">
              <a href="/tags/evolution">
                
                <span>evolution</span>
                
              </a>
            </li>
            
            <li class="dropdown-item my-0 py-0 mx-0 px-0">
              <a href="/tags/protein">
                
                <span>protein</span>
                
              </a>
            </li>
            
            <li class="dropdown-item my-0 py-0 mx-0 px-0">
              <a href="/tags/spider">
                
                <span>spider</span>
                
              </a>
            </li>
            
            <li class="dropdown-item my-0 py-0 mx-0 px-0">
              <a href="/tags/venom">
                
                <span>venom</span>
                
              </a>
            </li>
            
          </ul>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#publications_selected">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        

        
        
        
          
        

        <li class="nav-item">
          <a class="nav-link" href="/cv/cv_2018.pdf">
            
            <span>CV</span>
            
          </a>
        </li>

        
        

      

        

        

        
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  












  

  
  
  
<div class="article-container pt-3">
  <h1 itemprop="name">Reproducubility Tutorial</h1>

  
  <p class="page-subtitle">Tutorial for bioinformatics pipelines</p>
  

  
    

<div class="article-metadata">

  
  
  
  <div>
    <span itemscope itemprop="author" itemtype="http://schema.org/Person">
      <span itemprop="name">Jeff Cole</span>
    </span>
    
  </div>
  

  <span class="article-date">
    
    <meta content="2017-04-18 00:00:00 -0400 EDT" itemprop="datePublished">
    <time datetime="2017-04-18 00:00:00 -0400 EDT" itemprop="dateModified">
      Apr 18, 2017
    </time>
  </span>
  <span itemscope itemprop="publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="T. Jeffrey Cole">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    4 min read
  </span>
  

  
  

  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Reproducubility%20Tutorial&amp;url=%2fpost%2freproducibilitytutorial%2f"
         target="_blank" rel="noopener">
        <i class="fab fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2freproducibilitytutorial%2f"
         target="_blank" rel="noopener">
        <i class="fab fa-facebook-f"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2freproducibilitytutorial%2f&amp;title=Reproducubility%20Tutorial"
         target="_blank" rel="noopener">
        <i class="fab fa-linkedin-in"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2freproducibilitytutorial%2f&amp;title=Reproducubility%20Tutorial"
         target="_blank" rel="noopener">
        <i class="fab fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Reproducubility%20Tutorial&amp;body=%2fpost%2freproducibilitytutorial%2f">
        <i class="fas fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>

    







  









  
</div>



  <div class="article-container">

    <div class="article-style" itemprop="articleBody">
      

<h2 id="topics-to-be-covered">Topics to be covered</h2>

<ol>
<li>Anaconda</li>
<li>Snakemake</li>
<li>Docker</li>
</ol>

<h1 id="let-s-try-to-make-a-small-bioinformatics-pipeline">Let&rsquo;s try to make a small bioinformatics pipeline</h1>

<p>Let&rsquo;s say we have some FASTA files containing protein sequences, and we&rsquo;d like to make a phylogeny with those sequences.</p>

<p>To do this we would make the following bash script, called <code>butterbean.sh</code>
It would look like the following</p>

<pre><code class="language-bash">#!/bin/bash
mafft --auto --phylipout  $1 &gt; $1.aln
fasttree  $1.aln  &gt; $1.tree
</code></pre>

<p><strong>Data for the pipeline!</strong></p>

<p>Create a file called <code>ex1.fasta</code> from <a href="https://raw.githubusercontent.com/tijeco/ReproducibilityTutorial/master/ex1.fasta" target="_blank">here</a></p>

<h2 id="the-pipeline-works-the-following-way">The pipeline works the following way:</h2>

<p>First, <code>mafft</code> takes as input a fasta file of protein sequences as from input file the first argument <code>$1</code>, in this case <code>ex1.fasta</code>. The program then automatically chooses the best algorithm for generating an alignment as prompted by <code>--auto</code>, it then produces a multiple sequence alignment named <code>ex1.fasta.aln</code>  from <code>$1.aln</code>
 Then, <code>fasttree</code> takes as input the ouput from <code>mafft</code> referenced as <code>$1.aln</code> in our case known as <code>ex1.fasta.aln</code>. It then prodeces a newick format phylogeny of the sequences in the file referenced as <code>$1.tree</code> in our case ex1.fasta.tree</p>

<h2 id="running-our-pipeline">Running our pipeline</h2>

<p>Issue the folowing command to run the pipeline <code>butterbean.sh</code> on our data  <code>ex1.fasta</code></p>

<pre><code class="language-bash">bash butterbean.sh ex1.fasta
</code></pre>

<h2 id="what-did-the-output-look-like">What did the output look like?</h2>

<p>You may have received the following error message</p>

<pre><code class="language-bash">butterbean.sh: line 2: mafft: command not found
butterbean.sh: line 3: fasttree: command not found
</code></pre>

<h2 id="what-went-wrong">What went wrong????? 😭 😭 😭</h2>

<p>Well,<code>mafft</code> and <code>fasttree</code> are dependencies for <code>butterbean</code> and <strong>must</strong> be installed on your system in order to work properly.</p>

<p>If only there were a system in place for managing software packages and dependencies&hellip;.</p>

<p></br>
</br></p>

<p><img src="https://conda.io/docs/_images/conda_logo.svg" alt="alt text" /></p>

<p></br>
Software package dependency and environment management system.</p>

<p>Allows for easy version control!</p>

<h2 id="make-a-conda-environment-for-butterbean-s-dependencies">Make a conda environment for butterbean&rsquo;s dependencies</h2>

<p>Make a file called <code>butterbean.requirements.txt</code> with the following information in it</p>

<pre><code>mafft=7.310
fasttree=2.1.9
</code></pre>

<p>This file can be used to create a conda environment called <strong>{your_name_here}_butterbean</strong> using the following command</p>

<h3 id="do-not-use-upper-case-in-your-name-here">DO NOT USE UPPER CASE IN {your_name_here}</h3>

<pre><code class="language-bash">conda create --name {your_name_here}_butterbean -c bioconda --file butterbean.requirements.txt
</code></pre>

<p>Issue the following command to activate this environment</p>

<pre><code class="language-bash">source activate {your_name_here}_butterbean
</code></pre>

<p>Now we can rerun the pipeline</p>

<pre><code class="language-bash">bash butterbean.sh ex1.fasta
</code></pre>

<p>Hopefully it actually worked!</p>

<p>You should have a tree!</p>

<p><strong>What if we wanted to run our pipeline on hundreds of fasta files?</strong></p>

<p><strong>Can we add multithread support?</strong></p>

<p><strong>Can we have the dependencies automatically install themselves?</strong></p>

<h1 id="snakemake"><strong>Snakemake</strong></h1>

<p><img src="https://bitbucket-assetroot.s3.amazonaws.com/c/photos/2012/Dec/31/snakemake-logo-1807870285-12_avatar.png" alt="" /></p>

<p>This is a workflow management system that can increase the usability and reproducibility of our pipeline</p>

<h2 id="setup-conda-environments">Setup conda environments</h2>

<p>First, we need to take care of dependencies. We can set up conda environments for snakemake to setup and run on the fly during analysis</p>

<p>make a directory called <code>envs</code></p>

<p>In that directory, make the file <code>envs/mafft.yaml</code> with the following contents</p>

<pre><code class="language-yaml">channels:
  - bioconda
dependencies:
  - mafft=7.310
</code></pre>

<p>Make another file <code>envs/fasttree.yaml</code> with the following contents</p>

<pre><code class="language-yaml">channels:
  - bioconda
dependencies:
  - fasttree=2.1.9
</code></pre>

<h2 id="make-snakefile">Make Snakefile</h2>

<p>Make a file called <code>Snakefile</code> with the following information in it</p>

<pre><code class="language-python">SAMPLES, = glob_wildcards(&quot;{sample}.fasta&quot;)
</code></pre>

<p>This uses wild cards to find all files that end in <strong>.fasta</strong> that will be used by the pipeline</p>

<p>Add the final subroutine to <code>Snakefile</code></p>

<pre><code class="language-python">
rule final:
    input:expand(&quot;{sample}.tree&quot;,sample=SAMPLES)
</code></pre>

<p>Then, add the first subroutine</p>

<pre><code class="language-python">rule mafft:
    input:
        &quot;{sample}.fasta&quot;
    output:
        &quot;{sample}.aln&quot;
    conda:
        &quot;envs/mafft.yaml&quot;
    shell:
        &quot;mafft --auto --phylipout {input} &gt; {output}&quot;

</code></pre>

<p>Then, we add the second subroutine</p>

<pre><code class="language-python">rule fasttree:
    input:
        &quot;{sample}.aln&quot;
    output:
        &quot;{sample}.tree&quot;
    conda:  
        &quot;envs/fasttree.yaml&quot;
    shell:
        &quot;fasttree  {input}  &gt; {output}&quot;
</code></pre>

<p>Altogether, the file looks like the following</p>

<pre><code class="language-python">SAMPLES, = glob_wildcards(&quot;{sample}.fasta&quot;)

rule final:
    input:expand(&quot;{sample}.tree&quot;,sample=SAMPLES)

rule mafft:
    input:
        &quot;{sample}.fasta&quot;
    output:
        &quot;{sample}.aln&quot;
    conda:
        &quot;envs/mafft.yaml&quot;
    shell:
        &quot;mafft --auto --phylipout {input} &gt; {output}&quot;

rule fasttree:
    input:
        &quot;{sample}.aln&quot;
    output:
        &quot;{sample}.tree&quot;
    conda:
        &quot;envs/fasttree.yaml&quot;
    shell:
        &quot;fasttree  {input}  &gt; {output}&quot;


</code></pre>

<p>Run the pipeline with the following command</p>

<pre><code class="language-bash">snakemake --use-conda
</code></pre>

<p><strong>What if we want others to rerun our analysis on their computer with a non supported operating system?</strong></p>

<p><strong>Can we distribute our pipeline as a self contained virtual environment</strong></p>

<p><img src="https://www.docker.com/sites/default/files/social/docker_twitter_share.png" alt="" /></p>

<p>Create a file <code>Dockerfile</code> with the following information</p>

<pre><code class="language-Docker">FROM ubuntu
MAINTAINER Jeff Cole &lt;coleti16@students.ecu.edu&gt;


RUN apt-get -qq update
RUN apt-get install -y wget git build-essential cmake unzip curl
RUN apt-get install -qqy python3-setuptools python3-docutils python3-flask
RUN easy_install3 snakemake

RUN echo 'export PATH=/opt/conda/bin:$PATH' &gt; /etc/profile.d/conda.sh &amp;&amp; \
    wget --quiet https://repo.continuum.io/miniconda/Miniconda3-4.3.14-Linux-x86_64.sh -O ~/miniconda.sh &amp;&amp; \
    /bin/bash ~/miniconda.sh -b -p /opt/conda &amp;&amp; \
    rm ~/miniconda.sh

ENV PATH /opt/conda/bin:$PATH

WORKDIR /home/user/

RUN git clone https://github.com/tijeco/ReproducibilityTutorial.git

WORKDIR /home/user/ReproducibilityTutorial

RUN ln -sf /bin/bash /bin/sh

RUN snakemake --use-conda
</code></pre>

<p>Build the docker container</p>

<pre><code class="language-bash">docker build -t {your_name_here}_butterbean ./

</code></pre>

<p>Run the docker container with the following command</p>

<pre><code class="language-bash">docker run -it {your_name_here}_butterbean /bin/bash
</code></pre>

    </div>

    


<div class="article-tags">
  
  <a class="badge badge-light" href="/tags/tutorials/">tutorials</a>
  
</div>



    



  







    
    
    <div class="article-widget">
      <div class="hr-light"></div>
      <h3>Related</h3>
      <ul>
        
        <li><a href="/post/molecularevolutiontutorial/">Molecular Evolution Tutorial</a></li>
        
      </ul>
    </div>
    

    

    


  </div>
</article>

<div class="container">
  <footer class="site-footer">
  
  <p class="powered-by">
    <a href="/privacy/"></a>
  </p>
  

  <p class="powered-by">
    &copy; 2018 tijeco &middot; 

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" id="back_to_top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

</div>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha512-+NqPlbbtM1QqiK8ZAo4Yrj2c4lNQoGv8P79DPtKzj++l5jnN39rHA/xsqn8zE9l0uSoxaCdrOgFs6yjyfbBxSg==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha256-VsEqElsCHSGmnmHXGQzvoWjWwoznFSZc6hs7ARLRacQ=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
        
      

      
      
    

    
    

    
    
    

    
    
    <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script>
      const search_index_filename = "/index.json";
      const i18n = {
        'placeholder': "Search...",
        'results': "results found",
        'no_results': "No results found"
      };
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    

    

    
    

    
    

    
    
    
    
    
    
    
    
    <script src="/js/academic.min.74b0d3361475b08c40cf31120e82937b.js"></script>

    

  </body>
</html>

